firedrake/interpolation.py

```diff
@@ -381,6 +387,27 @@ def _interpolator(V, tensor, expr, subset, arguments, access):
         return copyin + (parloop_compute_callable, ) + copyout
 
 
+try:
+    _expr_cachedir = os.environ["FIREDRAKE_TSFC_KERNEL_CACHE_DIR"]
+except KeyError:
+    _expr_cachedir = os.path.join(tempfile.gettempdir(),
+                                  f"firedrake-tsfc-expression-kernel-cache-uid{os.getuid()}")
+
+
+def _compile_expression_key(expr, to_element, ufl_element, **kwargs):
+    try:
+        comm = kwargs["domain"].comm
+    except KeyError:
+        comm = expr.ufl_domains()[0].comm
+    key = hash(expr), to_element, ufl_element, utils.tuplify(kwargs.get("parameters"))
+    return comm, key
+
+
+compile_expression = disk_cached({}, _expr_cachedir, key=_compile_expression_key,
```

Comment:
Any reason you're not using `@disk_cached`? (I've never written a decorator)

Reply:
In Python a decorator is just a function that takes a function as its argument. It's hard to explain but what I have done here is equivalent to
```py
@disk_cached(...)
def compile_expression(*args, **kwargs):
    return compile_expression_dual_evaluation(*args, **kwargs)
```
just on fewer lines.

Reply:
The latter looks a bit nicer though IMO

Reply:
That's what I've ended up using anyway.